### PR TITLE
`Sps-Service` Header Proposal

### PR DESCRIPTION
**Pull request by nicholas-s-perkins**
Supporting Reusable Model Draft: https://github.com/SPSCommerce/sps-api-design/pull/161

### CODE REVIEW

File: standards/request-response.md [hunk 1]
```diff
@@ -527,6 +527,47 @@ Sps-Execution-Context:                  // values must be at least a character l
...
+what an API would rather know is not what "user" but what "client" is making the request.
+
+- `Sps-User-Agent` **MUST** be provided for all API requests that cannot include serviceName and serviceId in the `User-Agent`.
+- The header value **MUST** follow this format `serviceName[/serviceVersion] [sdk[/sdkVersion ](serviceTechRegistryId)`
+  - `^(?<NAME>[a-z0-9_-]{1,60})(?:/[^ ]+)? (?<SDK>[^ /]+(?:/[^ /]+)?)? ?\((?<SERVICEID>[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})\)$`
+- The header value **SHOULD** contain the SDK of the client.
+- It is expected to be rare that a service includes its version in the header, but it is an allowed option.
+- Requests with an invalid `Sps-User-Agent` **MUST** return a `400` response status code.
+- A serviceName **MAY** use `_` to delineate sub-clients of that service.
+- An API is expected to be able to parse User-Agent or this header for tech-registry and serviceName validation.
```
Comment (nicholas-s-perkins) [inline standards/request-response.md:553]: Including the regex for debate of the exact format.
I think it is important that it follows a very strict format to be useful, though.

File: standards/request-response.md [hunk 2]
```diff
@@ -527,6 +527,47 @@ Sps-Execution-Context:                  // values must be at least a character l
...
+**Description**: 
+`Sps-User-Agent` is a direct augmentation of `User-Agent` and is intended to be paired with it, following
+the format of the SPS [Authentication and Authorization Guardrail](https://atlassian.spscommerce.com/wiki/display/Guardrails/[Authentication+and+Authorization).
+Typically, an API should require the `User-Agent` header, and that it **SHOULD** follow the format 
+`serviceName[/serviceVersion] [sdk[/sdkVersion ](serviceTechRegistryId)` to include the valid tech-registry
+serviceName and serviceId of the client.
+However, is common for some clients, such as web-browsers, to lock a client out of setting the `User-Agent`.
+Thus, `User-Agent` cannot reliably encode actually identifying information for general applications.
+Furthermore, auth tokens also do not include this information, and it is also common to forward a user_token, where
+what an API would rather know is not what "user" but what "client" is making the request.
```
Comment (nicholas-s-perkins) [inline standards/request-response.md:544]: Could be a bit wordy, but this is my attempt to explain the long causality that causes one to want this header in the critical API space. 

parcel-service and NCS already support a custom header like this, which is `X-User-Agent-Override`, and we find it very valuable for understanding UI behavior.

One thing this doesn't quite call out is that I think any service that implements this header then should require the correct format for `User-Agent` as well.  It's more like a cascade.
1. Check `Sps-User-Agent` first, set as Agent value.
2. If `Sps-User-Agent` not present, set `User-Agent` as Agent value.
3. If Agent value is invalid format, reject with a `400`

Just some higher level thought about this, but there's also opportunities here possibly for tooling like atlas service mesh to populate this for some requests to give apps a better automatic user-agent.

File: standards/request-response.md [hunk 3]
```diff
@@ -527,6 +527,47 @@ Sps-Execution-Context:                  // values must be at least a character l
 Sps-Execution-Context: 1                // valid, but SHOULD be human-readable.
 ```
 
+#### Sps-User-Agent
```
Comment (travisgosselin) [inline standards/request-response.md:530]: User-Agent traditionally identifies the user-facing client, such as a browser or device. Service-to-service communication tracking is a distinct concept in my opinion, and conflating the two can lead to confusion. A unique header name clarifies its purpose, making it easier to explain, document, and enforce. I love this concept and the details, but I think we should workshop a different name for the header.   (SPS-Service, SPS-Origin-Service, SPS-Service-Client, etc)

Comment (JamesStauffer) [reply standards/request-response.md:530]: Should it be SPS specific? If an external client uses an API through a browser, I assume we would still want it provided.

Comment (nicholas-s-perkins) [reply standards/request-response.md:530]: > Service-to-service communication tracking is a distinct concept in my opinion, and conflating the two can lead to confusion.
It could be, but our guardrails already pushes apps to treat `User-Agent` as the place to put serviceName/serviceId, and this is really just trying to conceive of a way to support that for web-uis (which, we don't always think of as a 'service' but for all intents and purposes of this they are).  The format for it is right out of our guardrails for User-Agent.  
As a point of reference, we already require teams to use a good user-agent for parcel-service, but we can't enforce it, so it's an immense amount of work to track down new users that refuse to follow our guidelines and have them update their agents.
What I actually wanted to do was just force the regex on User-Agent itself for any new API we make, but Chrome hates us.  So the squaring of that is to enforce the regex on both User-Agent or Sps-User-Agent, where `Sps-User-Agent` merely exists for things that cannot set a good User-Agent.

I feel like the naming denotes how they are essentially the exact same concept, it's not a conflation of concepts they are intended to be the same concept.  But, I'm not super wedded to it.  Practically, most users are going to find out about what the headeris called because my API rejects their web requests.

Comment (travisgosselin) [reply standards/request-response.md:530]: I can see different reasons where the continued usage of User-Agent makes sense as is to denote client details about the "instance" make the request, such as a browser (browser, web app), python version (backend), etc. Using something more structured and designated for service metadata allows for the operation of the normal user agent to continue without confusion for "SPS-User-Agent". Additionally, I think the service metadata implies something more static about the application consumer (not the user or context of the user). 

I'd advocate for something specifically for this purpose: 
SPS-Service: xxxx-xxxx-xxxxx-xxxx (my-service-name) 

This data as an exact term can be filled in via other infrastructure in the future as well, such as Atlas egress.  Thoughts?

Comment (nicholas-s-perkins) [reply standards/request-response.md:530]: As I've thought about, I'm really just trying to solve a problem with wanting service-name + service-id.
I'm somewhat agnostic as to the exact approach, but I do want to require it in some way for my new service since I do really want to solve the ambiguous caller problem that we have in our network.
If we want a more purpose built header that I can require for all requests, that's also fine.
Freeing ourselves from user-agent, it might be interesting to shift user-agent as more to the "what is this thing" question.  Like it's more what tells you if something is a browser, or a java/pythong thing.

Then my app would have to require another header (or headers?).

One simplification could be two headers:
```
Sps-Service-Name: my-service
Sps-Service-Id: d8289b6e-5af7-4568-a1be-ebdf872b2c44
```
Certainly would make it easier to parse and there would be no ambiguity about the format, but also that's kind of a bummer to require two headers like this.

Combining them into your idea of `Sps-Service` is a bit harder to parse, but probably more ergonomic for clients. Then you have to debate format a bit.  A lot of ways you could slice it, but space separated is probably the most simple (?) and we don't allow them for name nor ids.  It also matches how user-agent separates values and is trivial to parse.  I don't really care what the order is, but I might vote for `{name} {id}` as it reads a bit more naturally to the eye to me and matches what was our user-agent order guideline:
```
Sps-Service: my-service d8289b6e-5af7-4568-a1be-ebdf872b2c44
Sps-Service: d8289b6e-5af7-4568-a1be-ebdf872b2c44 my-service
```
Worth noting, we did come up with a nomenclature in some of the clients called the "app agent", which we could use instead but that was in part to align with injecting the values into User-Agent and kind of matching the naming that it was an "Agent".

So `Sps-Service` is maybe the best option, so I'd also vote for that.  I also don't think it prevents a future of if `Sps-Service-Id` becomes a thing at a later point, and could work as a sugar for it.

Comment (travisgosselin) [reply standards/request-response.md:530]: > Like it's more what tells you if something is a browser or a java/python thing. In that design lens, then, my app might not even care if something has a User-Agent for the most part.

My intent, perspective is to try and divorce this from being a user-agent solution/problem. That aside, user-agent is required in our API Standards today, and is enforced header you have to pass on all Network API requests as part of the incoming gateway.

> Certainly would make it easier to parse and there would be no ambiguity about the format, but also that's kind of a bummer to require two headers like this.

Agreed, I think we can lean on a single format with a single header for now that combines the name and ID. That being said, can you provide more context on the need for "name"? It is something that is not immutable / can be changed in tech registry usage. I know its convenient to have, are there other implications of not having it? Does parcel service require the name today? Should name be optional, while service id required?

Comment (nicholas-s-perkins) [reply standards/request-response.md:530]: > That being said, can you provide more context on the need for "name"? It is something that is not immutable / can be changed in tech registry usage. I know its convenient to have, are there other implications of not having it? Does parcel service require the name today? Should name be optional, while service id required?

For `Sps-Service` the name should be required, and the format of the header should not have more than one fixed pattern.  I think a future expansion to only including the ID should probably be reserved for using the `Sps-Service-Id` header, separately.

On top of providing a serviceId, you can think of this header an audit string for the caller (with the reminder that the identity auth audit string does not provide any caller information whatsoever).  So the utility of the name is primarily for logs and debugging.
I think it's reasonable to have this format and allow the name to be fungible, though it could get called out in the docs.  Essentially the name gives log friendly context to the ID, where the ID is what should be used for any programmatic associations.

In example, if we take a look at a current parcel-service access log:
```json
{
  "timestamp": "2024-12-26T18:55:52.115Z",
  "contentLength": 0,
  "method": "PUT",
  "protocol": "HTTP/1.1",
  "remoteAddress": "127.0.0.6",
  "remoteUser": "APIKey 665 Org 157627141979612915626333577105111854232 (SPS Commerce)",
  "requestTime": 254,
  "uri": "/v3/parcels/55904180953/ackinfo",
  "status": 204,
  "userAgent": "acknowledgementservice/8.3.1 Java/17.0.4",
  "forwardedFor": "10.142.4.50",
  "forwardedHost": "parcel.api.spsprod.in",
  "context": "prod",
  "requestId": "76e42882-1a5c-4fec-8b42-6618e0b11ade"
}
```
The most important field on that log for developers is the User-Agent, it's the only thing that really identifies what this is in a coherent way (reminding that `APIKey 665` is not bound to any service nor team, and also there's no developer facing API that you can tell what APIKey 665 even is):
```
"userAgent": "acknowledgementservice/8.3.1 Java/17.0.4",
```
But this doesn't work that well, of course, because there's no enforcement of this user-agent.  You can note that there isn't even a serviceId in this example.
If we imagine our future payload-storage-service audit log, we would seek to replace (or augment) userAgent with `Sps-Service`.
```
{
  "method": "POST",
  "remoteUser": "APIKey 665 Org 157627141979612915626333577105111854232 (SPS Commerce)",
  "uri": "/v1/payloads",
  "status": 201,
  "userAgent": "whatever",
  "spsService": "acknowledgementservice 1c3857ee-8fbe-4763-b9f4-fe6a62f697b4",
  "requestId": "76e42882-1a5c-4fec-8b42-6618e0b11ade"
}
```
` "spsService": "acknowledgementservice 1c3857ee-8fbe-4763-b9f4-fe6a62f697b4"` is 🔥 🔥 🔥 
It's a huge benefit for very little cost.

**Review by jimthedev — COMMENTED** [review]
Hello. Thanks for working on this.

Just a few (perhaps naïve) questions:

- Would it be fair to assume that the id & name portion of the Sps-Service header should be registered in prod tech registry? If so, would it be worth mentioning in the notes that this as a place where people should go to get that service id and name to at least try to ensure they match?
- Would the environment of the service making the request be something that might be added into the custom attributes given that it isn't part of the key itself? Perhaps that is actually somewhere else in the spec or maybe is being considered in other specs. I am asking because it at least is possible that in the case of a public endpoint it could get a request from another environment than the one the receiving service is running in. This kind of thing would be tricky to track down but would result in prod data being returned to a test UI or vice versa. Maybe not something for the spec but something to think about when considering what we could log given the header.

Again thanks for doing this.

**Comment by travisgosselin** [comment]
> Hello. Thanks for working on this.
> 
> Just a few (perhaps naïve) questions:

Great feedback @jimthedev  - appreciate you taking the time to read through it.

> Would it be fair to assume that the id & name portion of the Sps-Service header should be registered in prod tech registry? If so, would it be worth mentioning in the notes that this as a place where people should go to get that service id and name to at least try to ensure they match?

Our API Standards ideally are created agnostic of the implementation and subsystem that sits on top of it and references it. They are also public (like this conversation). So with that in mind,  it is recognized that there needs to be an authority for the service ID (in our case at SPS that is tech registry). This line was intending to cover those use cases: `Requests **MAY** validate the `Sps-Service` header `id` value against the associated authority or service registry. Validation failures **MUST** return a 400 indicating the reason.`

Does that make sense or should we add more / reword?

> Would the environment of the service making the request be something that might be added into the custom attributes given that it isn't part of the key itself? Perhaps that is actually somewhere else in the spec or maybe is being considered in other specs. I am asking because it at least is possible that in the case of a public endpoint it could get a request from another environment than the one the receiving service is running in. This kind of thing would be tricky to track down but would result in prod data being returned to a test UI or vice versa. Maybe not something for the spec but something to think about when considering what we could log given the header.

I think of this header as environment agnostic, and specific to the service entity itself. With that in mind, the implementation architecture matters to ensure it is understood what environment the service is from. That being said, the state in the SPS ecosystem is that we have one PROD asset registry used in all environments, our runtime in our integration environment makes use of PRODUCTION service ids as a single source of truth for a lot of reasons I won't enumerate here. Please advise me if you think this is incorrect.

**Review by acchristensen — APPROVED** [review]
Looking forward to having this in and folks adopting it!

**Comment by jimthedev** [comment]
@travisgosselin Ok I think that makes sense. Thanks for clarifying. I think that rewording makes a lot of sense.

**Comment by travisgosselin** [comment]
> @travisgosselin Ok I think that makes sense. Thanks for clarifying. I think that rewording makes a lot of sense.

Any suggestions on how to reword it? What terminology you'd like to see to make it clear? I put it through AI to try and get some alternatives, but mostly just a few grammatical changes:

```
Requests MAY validate the Sps-ServiceHeaderId value against the corresponding authority or service registry. If validation fails, a 400 Bad Request response MUST be returned with an explanation of the error.
``

Thoughts @jimthedev ?
